### PR TITLE
Add `deterministic?` option to force consistent ->Javascript ordering

### PR DESCRIPTION
**Pull request by sritchie**
This bit me in the double pendulum tests. I didn't change the default behavior, but added an option that lets us control the variable order.

### CODE REVIEW

File: src/sicmutils/numerical/compile.cljc [hunk 1]
```diff
@@ -69,33 +69,40 @@
...
+      (let [cs (atom {})
+            increment (fnil inc 0)]
+        ;; cs maps subexpressions to the number of times we have seen the
+        ;; expression.
+        (w/postwalk (fn [e]
+                      (when (and (seq? e) (not (expr-to-var e)))
+                        (swap! cs update e increment))
+                      e)
+                    x)
+        (let [new-syms (into {} (for [[k v] (pairs @cs) :when (> v 1)] [k (symbol-generator)]))]
```
Comment (sritchie) [inline src/sicmutils/numerical/compile.cljc:99]: This is the change - `@cs` becomes `(pairs @cs)`, so we can optionally control the sort order. We sort the string representations, here.

File: src/sicmutils/numerical/compile.cljc [hunk 2]
```diff
@@ -105,22 +112,28 @@
...
-        new-expression))))
+   x
+   symbol-generator
+   (fn [new-expression new-vars]
+     (if (> (count new-vars) 0)
+       (do
+         (log/info (format "common subexpression elimination: %d expressions" (count new-vars)))
+         `(let ~(initialize-cs-variables new-vars) ~new-expression))
+       new-expression))
+   :deterministic? deterministic?))
```
Comment (sritchie) [inline src/sicmutils/numerical/compile.cljc:136]: I let the option pass through in a few spots.

File: test/sicmutils/infix_test.cljc [hunk 3]
```diff
@@ -240,32 +240,18 @@
...
+        (is (= (format
+                (str "function(dx, dy, f, partial, x, y) {\n"
+                     "  var _0003 = partial(0);\n"
+                     "  var _0004 = partial(1);\n"
+                     "  var _0005 = [x, y];\n"
+                     "  var _0006 = _0003(f);\n"
+                     "  var _0007 = _0004(f);\n"
+                     "  return %s * Math.pow(dx, 2) * _0003(_0006)(_0005) + dx * dy * _0004(_0006)(_0005) + %s * Math.pow(dy, 2) * _0004(_0007)(_0005) + dx * _0006(_0005) + dy * _0007(_0005) + f(_0005);\n}")
+                half half)
+               (s->JS expr :deterministic? true)))
```
Comment (sritchie) [inline test/sicmutils/infix_test.cljc:254]: woohoo!

**Comment by littleredcomputer** [comment]
very sweet. trying to understand how variable order worked, and what purpose it was attempting to achieve, in the original code base was a sore challenge.

**Comment by sritchie** [comment]
Whoops, closing since I merged #85 , which contains this already.

**Pull request closed without merging**